### ISSUE TITLE
How can you tell markid to ignore some identifiers?

### ISSUE THREAD

**Issue by Jacob-Flasheye** [David-Kunz/markid#5]
Hi!

I really like the idea behind this plugin, but there are some identifiers that I do not want it to highlight, for example the identifier in between the `from` and `import` in this python code:
```python
from foo.bar import baz
```
How would I instruct markid to ignore identifiers depending on their place in the tree?

**Comment by David-Kunz** [comment]
Hi @Jacob-Flasheye ,

Thanks a lot!

In principle, you can create a custom Tree-Sitter query for identifier selection and set it in `queries.default` (or for your specific language).

Hint: You can use the tool [Tree-Sitter Playground](https://github.com/nvim-treesitter/playground) to play around with queries.

Best regards,
David

**Comment by ziontee113** [comment]
I use this query to ignore any `require` or `ipairs` identifiers in Lua
```lua
markit.queries = {
	default = '((identifier) @markid (#not-any-of? @markid "require" "ipairs"))',
}
```

**Comment by David-Kunz** [comment]
Hi @ziontee113 ,

Unfortunately I haven't played around with exclusion queries, but you can try it with [treesitter-playground](https://github.com/nvim-treesitter/playground).

Maybe with [negated fields](https://tree-sitter.github.io/tree-sitter/using-parsers#query-syntax)?

> Negated Fields
You can also constrain a pattern so that it only matches nodes that lack a certain field. To do this, add a field name prefixed by a ! within the parent pattern. For example, this pattern would match a class declaration with no type parameters:

```

(class_declaration
  name: (identifier) @class_name
  !type_parameters)
```

**Comment by David-Kunz** [comment]
@ziontee113 , I see.

I'd rather not include that in markid as this would mean to not only retrieve all matches for the first query, but also for the exclusion query. This could cripple the performance.

**Comment by ziontee113** [comment]
> I'd rather not include that in markid as this would mean to not only retrieve all matches for the first query, but also for the exclusion query and then comparing both. This could cripple the performance.

I agree. Thank you very much for the amazing plugin :heart:

**Comment by David-Kunz** [comment]
Thank you, @ziontee113 !

**Comment by igorgue** [comment]
@David-Kunz I was trying to ignore the params highlight and I found of a useful `[not]-has_parent` predicate:

```lua
queries = {
    default = '((identifier) @markid (#not-has-parent? @markid parameters typed_parameter typed_default_parameter))',
},
```

But doesn't really `follow` the identifier and it keeps getting colored, look:

![image](https://user-images.githubusercontent.com/7014/199817356-7741c1a9-d2cb-432e-ada4-42f49a49a61f.png)

**Comment by David-Kunz** [comment]
Thanks for the info @igorgue . What happens if you run this query in TSPlatground, are the right identifiers highlighted?

**Comment by ziontee113** [comment]
Hi @David-Kunz , @igorgue , I can confirm this query works on Treesitter Playground and with Markid:
```
( (identifier) @markid (#not-has-ancestor? @markid jsx_expression jsx_opening_element jsx_closing_element) )
```
or
```
( (identifier) @markid (#not-has-parent? @markid jsx_expression jsx_opening_element jsx_closing_element) )
```

**Comment by igorgue** [comment]
@David-Kunz It's okay I just wanted to help, but no, the query would get the identifier in the params but it'd still get selected since it's an identifier in the block of the function... I'm still thinking about the query

**Comment by igorgue** [comment]
@David-Kunz maybe like `Formatter` the plugin [1], you can have a way to have default `queries` by language in the repo, since I think most people I show this they say it's too much to color every identifier, I think is a great default, but I've seen the opinion, I just don't know which those defaults are, on the top of my mind probably do not color constants?

https://github.com/mhartington/formatter.nvim/tree/master/lua/formatter/filetypes

**Comment by David-Kunz** [comment]
That's true @igorgue , for many it's too colourful. It's just hard to come up with better defaults without introducing many language-specific queries. I haven't yet come to a solution.
